ethereum/bridge/contracts/Bridge.sol

```diff
@@ -0,0 +1,295 @@
+
+	// Function to pause the bridge
+	function resumeBridge() private isPaused {
+		paused = false;
+	}
+
+	// modifier to check if bridge is running
+	modifier isRunning() {
+		// If the first argument of 'require' evaluates to 'false', execution terminates and all
+		// changes to the state and to Ether balances are reverted.
+		// This used to consume all gas in old EVM versions, but not anymore.
+		// It is often a good idea to use 'require' to check if functions are called correctly.
+		// As a second argument, you can also provide an explanation about what went wrong.
+		require(paused == false, 'Bridge is not Running');
+		_;
+	}
+
+	// modifier to check if bridge is paused
+	modifier isPaused() {
+		// If the first argument of 'require' evaluates to 'false', execution terminates and all
+		// changes to the state and to Ether balances are reverted.
+		// This used to consume all gas in old EVM versions, but not anymore.
+		// It is often a good idea to use 'require' to check if functions are called correctly.
+		// As a second argument, you can also provide an explanation about what went wrong.
+		require(paused == true, 'Bridge is Paused');
+		_;
+	}
+
+	// Event to emit when a transfer is initiated
+	event ValidatorAdded(
+		address addr, // The address of the validator
+		uint256 weight // The weight of the validator
+	);
+
+	function initialize(Validator[] calldata _validators) public initializer {
+		// addValidator(firstPK, firstWeight);
+		// __Ownable_init();
+		__UUPSUpgradeable_init();
+		paused = false;
+
+		for (uint256 i = 0; i < _validators.length; i++) {
+			addValidator(_validators[i].addr, _validators[i].weight);
+			emit ValidatorAdded(_validators[i].addr, _validators[i].weight);
+		}
+	}
+
+	// constructor() {
+	//     _disableInitializers();
+	// }
+
+	// /// @custom:oz-upgrades-unsafe-allow constructor
+	// constructor() initializer {}
+
+	function testBridgeMessage(
+		BridgeMessage calldata bridgeMessage
+	) public pure returns (uint8, uint8, ChainID, uint64, address, uint8, address) {
+		return (
+			bridgeMessage.messageType,
+			bridgeMessage.version,
+			bridgeMessage.sourceChain,
+			bridgeMessage.bridgeSeqNum,
+			bridgeMessage.senderAddress,
+			bridgeMessage.targetChain,
+			bridgeMessage.targetAddress
+		);
+	}
+
+	function approveBridgeMessage(
+		BridgeMessage calldata bridgeMessage,
+		bytes[] calldata signatures
+	) public isRunning returns (bool, uint256) {
+		uint256 totalWeight = 0;
+		// verify signatures
+		bytes32 hash = ethSignedMessageHash(bridgeMessage);
+		for (uint256 i = 0; i < signatures.length; i++) {
+			address recoveredPK = recoverSigner(hash, signatures[i]);
+			// Check if the address is not zero
+			require(recoveredPK != address(0), 'Zero address.');
+			uint256 index = validatorIndex[recoveredPK] - 1;
+			require(index < validators.length, 'Index out of bounds');
+
+			Validator memory validator = validators[index];
+			require(recoveredPK == validator.addr, 'Invalid signature');
```

Comment:
can we have `mapping(address => Validator) public validators;` instead of a `Validator[]` and an index mapping?

Reply:
`mapping` doesn't allow the iteration of its entries, which is probably why we are doing this. But I think it's better to either write ourselves or use an existing an iterable mapping library that abstracts these operations away.